### PR TITLE
Screen Space and World Space Instanced Sprites (basic rendering), Render Pass Cycling, Orthographic Camera

### PR DESCRIPTION
**Pull request by wejdell**

### CODE REVIEW

File: Havtorn/Source/Editor/Windows/InspectorWindow.cpp [hunk 1]
```diff
@@ -110,6 +110,8 @@ namespace ImGui
 	{
 		if (selectedEntity->HasComponent(componentType))
 		{
+			// TODO.NR/AG: Extract Collapsing Header and Remove Component button logic and move it here
```
Comment (AxelSavage) [inline Havtorn/Source/Editor/Windows/InspectorWindow.cpp]: Just leaving a note from Discord discussion,
We can use GetComponentTypeString here to get the 'const char* label' parameter for ImGui::CollapsingHeader. Although we'll have to decide if we want the "component" part of the string or not :)

File: Havtorn/Source/Editor/Windows/InspectorWindow.cpp [hunk 2]
```diff
@@ -131,6 +133,9 @@ namespace ImGui
 		ImGui::DragFloat3("Rotation", matrixRotation, SlideSpeed);
 		ImGui::DragFloat3("Scale", matrixScale, SlideSpeed);
 		ImGuizmo::RecomposeMatrixFromComponents(matrixTranslation, matrixRotation, matrixScale, transformMatrix.data);
+
+		// TODO.NR: Fix rotation here, can't rotate full range of Yaw due to weird clamping. Other axes are fine
```
Comment (AxelSavage) [inline Havtorn/Source/Editor/Windows/InspectorWindow.cpp]: Found this Git-Issue from ImGuizmo's repo regarding this specific issue. Or I believe it's the same or similar issue!
https://github.com/CedricGuillemet/ImGuizmo/issues/244

File: Havtorn/Source/Engine/ECS/Components/SpriteComponent.h
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/ECS/Components/SpriteComponent.h:1]: How do we fix the "// Copyright .. " from being pasted twice?

File: Havtorn/Source/Engine/Graphics/RenderCommand.h [hunk 3]
```diff
@@ -22,19 +22,22 @@ namespace Havtorn
...
 		DeferredLightingPoint,
 		DeferredLightingSpot,
+		PostBaseLightingPass,
 		VolumetricLightingDirectional,
 		VolumetricLightingPoint,
 		VolumetricLightingSpot,
 		VolumetricBufferBlurPass,
 		ForwardTransparency,
-		OutlineMask,
-		Outline,
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/RenderCommand.h:37]: Why are these commented? Could you add a comment explaining why (to avoid confusion later) or remove them?
(Referring to //OutlineMask & //Outline)

File: Havtorn/Source/Engine/Graphics/Renderers/FullscreenRenderer.cpp [hunk 4]
```diff
@@ -57,35 +54,29 @@ namespace Havtorn
...
-		filepaths[static_cast<size_t>(EFullscreenShader::DepthAwareUpsampling)] = "Shaders/FullscreenDepthAwareUpsample_PS.cso";
-		//filepaths[static_cast<size_t>(EFullscreenShader::DeferredAlbedo)] = "Shaders/DeferredRenderPassShader_Albedo.cso";
-		//filepaths[static_cast<size_t>(EFullscreenShader::DeferredNormals)] = "Shaders/DeferredRenderPassShader_Normal.cso";
-		//filepaths[static_cast<size_t>(EFullscreenShader::DeferredRoughness)] = "Shaders/DeferredRenderPassShader_Roughness.cso";
-		//filepaths[static_cast<size_t>(EFullscreenShader::DeferredMetalness)] = "Shaders/DeferredRenderPassShader_Metalness.cso";
-		//filepaths[static_cast<size_t>(EFullscreenShader::DeferredAmbientOcclusion)] = "Shaders/DeferredRenderPassShader_AO.cso";
-		//filepaths[static_cast<size_t>(EFullscreenShader::DeferredEmissive)] = "Shaders/DeferredRenderPassShader_Emissive.cso";
-
-		for (UINT i = 0; i < static_cast<size_t>(EFullscreenShader::Count); i++) 
+		std::array<std::string, STATIC_U64(EFullscreenShader::Count)> filepaths;
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Renderers/FullscreenRenderer.cpp:57]: A suggestion:
Envelop all filepaths[] = in curly braces, for readability. Should also allow the scope to be collapsed.
I.e
```
std::array<std::string, STATIC_U64(EFullscreenShader::Count)> filepaths;
{
        filepaths[STATIC_U64(EFullscreenShader::Multiply)] = "Shaders/FullscreenMultiply_PS.cso";
        filepaths[STATIC_U64(EFullscreenShader::Copy)] = "Shaders/FullscreenCopy_PS.cso";
	filepaths[STATIC_U64(EFullscreenShader::CopyDepth)] = "Shaders/FullscreenCopyDepth_PS.cso";
	filepaths[STATIC_U64(EFullscreenShader::CopyGBuffer)] = "Shaders/FullscreenCopyGBuffer_PS.cso";
        ...
}
```

File: Havtorn/Source/Engine/Graphics/Shaders/FullscreenBilateralBlurHorizontal_PS.hlsl [hunk 5]
```diff
@@ -10,14 +10,14 @@ PixelOutput main(VertexToPixel input)
     float3 blurColor = float3(0.0f, 0.0f, 0.0f);
     float normalizationFactor = 0.0f;
     const float bZ = 1.0 / normpdf(0.0, BSIGMA);
-    const float3 originalPixelValue = fullscreenTexture1.Sample(defaultSampler, input.myUV.xy).rgb;
+    const float3 originalPixelValue = fullscreenTexture1.Sample(defaultSampler, input.UV.xy).rgb;
     //float3 originalDepth = fullscreenTexture2.Sample(defaultSampler, input.myUV.xy).rgb;
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Shaders/FullscreenBilateralBlurHorizontal_PS.hlsl]: Can the comments be removed?

File: Havtorn/Source/Engine/Graphics/Shaders/FullscreenBilateralBlurVertical_PS.hlsl [hunk 6]
```diff
@@ -10,14 +10,14 @@ PixelOutput main(VertexToPixel input)
     float3 blurColor = float3(0.0f, 0.0f, 0.0f);
     float normalizationFactor = 0.0f;
     const float bZ = 1.0 / normpdf(0.0, BSIGMA);
-    const float3 originalPixelValue = fullscreenTexture1.Sample(defaultSampler, input.myUV.xy).rgb;
+    const float3 originalPixelValue = fullscreenTexture1.Sample(defaultSampler, input.UV.xy).rgb;
     //float3 originalDepth = fullscreenTexture2.Sample(defaultSampler, input.myUV.xy).rgb;
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Shaders/FullscreenBilateralBlurVertical_PS.hlsl]: Ok remove comments?

File: Havtorn/Source/Engine/Graphics/Shaders/FullscreenCopyDepth_PS.hlsl [hunk 7]
```diff
@@ -21,11 +21,11 @@ float Remap(float inMin, float inMax, float outMin, float outMax, float v)
 PixelOutput main(VertexToPixel input)
 {
     PixelOutput returnValue;
-    const float resource = fullscreenTexture1.Sample(defaultSampler, input.myUV.xy).r;
+    const float resource = fullscreenTexture1.Sample(defaultSampler, input.UV.xy).r;
 
     //returnValue.myColor.r = Remap(0.99f, 1.0f, 0.0f, 1.0f, resource);
-    returnValue.myColor.r = resource.r;
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Shaders/FullscreenCopyDepth_PS.hlsl:27]: Can the comment be removed? Sorry for adding this in so many places.

File: Havtorn/Source/Engine/Graphics/Shaders/FullscreenDepthDownSample_PS.hlsl [hunk 8]
```diff
@@ -7,12 +7,12 @@ PixelOutput main(VertexToPixel input)
     PixelOutput returnValue;
     //float4 depth2x2 = fullscreenTexture1.Load(float3(input.myUV.xy, 0));
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Shaders/FullscreenDepthDownSample_PS.hlsl]: Can the comments be removed? Or are they kept for a reason?

File: Havtorn/Source/Engine/Graphics/Shaders/FullscreenDifference_PS.hlsl [hunk 9]
```diff
@@ -0,0 +1,15 @@
+// Copyright 2022 Team Havtorn. All Rights Reserved.
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Shaders/FullscreenDifference_PS.hlsl:1]: Maybe we have to manually remove the 2022 comments?

File: Havtorn/Source/Engine/Graphics/Shaders/FullscreenVignette_PS.hlsl [hunk 10]
```diff
@@ -20,13 +20,13 @@ PixelOutput main(VertexToPixel input)
     //resource.y *= clamp(vignette + myVignetteColor.g, 0.0f, 1.0f);
     //resource.z *= clamp(vignette + myVignetteColor.b, 0.0f, 1.0f);
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Shaders/FullscreenVignette_PS.hlsl]: Sorry. Comments?

File: Havtorn/Source/Engine/Graphics/Shaders/SpriteInstanced_VS.hlsl [hunk 11]
```diff
@@ -0,0 +1,14 @@
...
+
+#include "Includes/SpriteShaderStructs.hlsli"
+
+InstancedVertexToGeometry main(InstancedVertexInput input)
+{
+    InstancedVertexToGeometry output;
+    output.Transform = input.Transform;
+    output.Color = input.Color;
+    output.UVRect = input.UVRect;
+	return output;
```
Comment (Aki-Gonzalez) [inline Havtorn/Source/Engine/Graphics/Shaders/SpriteInstanced_VS.hlsl:13]: The indentation on ``return output``, intended or ?

File: Havtorn/Source/Engine/ECS/Components/SpriteComponent.h
Comment (wejdell) [reply Havtorn/Source/Engine/ECS/Components/SpriteComponent.h:1]: We fix the batch script, I tried giving it a shot but the syntax is too hard to parse for me still. It's on the Todo-list! Will go through all files and fix them up then. We need to fix it anyway to enable ones with long path names to run it without it failing the command, but we could skip the copyright year for now. Should have a talk about it.

File: Havtorn/Source/Engine/Graphics/Shaders/FullscreenDifference_PS.hlsl (hunk 9 above)
Comment (wejdell) [reply Havtorn/Source/Engine/Graphics/Shaders/FullscreenDifference_PS.hlsl:1]: See https://github.com/nicolas-risberg/Havtorn/pull/19#discussion_r1281135540

File: Havtorn/Source/Engine/Graphics/Shaders/SpriteInstanced_VS.hlsl (hunk 11 above)
Comment (wejdell) [reply Havtorn/Source/Engine/Graphics/Shaders/SpriteInstanced_VS.hlsl:13]: There is no indentation on my end, maybe check it when it's merged? or if you check out my branch and can be bothered 🤷

File: Havtorn/Source/Engine/ECS/Components/CameraComponent.h
Comment (AxelSavage) [inline Havtorn/Source/Engine/ECS/Components/CameraComponent.h:1]: (Low prio and not neccesarily relevant to the pullrequest in general!)
Could be useful to move "camera settings" into a struct, just to make life easier.
We'll most likely switch alot between an Editor Camera & InGame Camera. These two most likely will exist and we'll want to switch between! Actually... Now when I think about it, SCameraComponent IS already this specific data! blah my b! but still sharing this to highlight my in-progress learning/understandind of ECS xD

File: Havtorn/Source/Engine/ECS/Systems/RenderSystem.cpp [hunk 12]
```diff
@@ -32,6 +32,10 @@ namespace Havtorn
 		const auto& decalComponents = scene->GetDecalComponents();
 
 		RenderManager->ClearSystemStaticMeshInstanceTransforms();
+		RenderManager->ClearSpriteInstanceWSTransforms();
+		RenderManager->ClearSpriteInstanceSSTransforms();
```
Comment (AxelSavage) [inline Havtorn/Source/Engine/ECS/Systems/RenderSystem.cpp]: Opinion: These might as well be worded out as WorldSpace & ScreenSpace (instead of WS / SS) <3

File: Havtorn/Source/Engine/ECS/Systems/RenderSystem.cpp [hunk 13]
```diff
@@ -238,19 +248,45 @@ namespace Havtorn
...
+				if (!RenderManager->IsSpriteInInstancedWSTransformRenderList(spriteComp.TextureIndex)) 
+				{
+					// NR: Don't push a command every time
+					SRenderCommand command;
+					command.SpriteComponent = spriteComp;
+					command.Type = ERenderCommandType::GBufferSpriteInstanced;
+					RenderManager->PushRenderCommand(command);
+				}
+
+				RenderManager->AddSpriteToInstancedWSTransformRenderList(spriteComp.TextureIndex, transformComp.Transform.GetMatrix());
```
Comment (AxelSavage) [inline Havtorn/Source/Engine/ECS/Systems/RenderSystem.cpp]: In reference to my previous comment on line 36.
Hmmm Maybe it's a bit too much? <3

```suggestion
				RenderManager->AddSpriteToInstancedWorldSpaceTransformRenderList(spriteComp.TextureIndex, transformComp.Transform.GetMatrix());
```

File: Havtorn/Source/Engine/Graphics/GraphicsEnums.h [hunk 14]
```diff
@@ -24,7 +24,8 @@ namespace Havtorn
 	{
 		Pos3Nor3Tan3Bit3UV2,
 		Pos3Nor3Tan3Bit3UV2Trans,
-		Pos4
+		Pos4,
```
Comment (AxelSavage) [inline Havtorn/Source/Engine/Graphics/GraphicsEnums.h]: The other members here are clearly defined, maybe this Pos4 member could use a teeny tiny bit more love?

File: Havtorn/Source/Engine/Graphics/RenderManager.cpp [hunk 15]
```diff
@@ -141,6 +87,40 @@ namespace Havtorn
 		InitEditorResources();
 		LoadDemoSceneResources();
 
+		RenderFunctions[ERenderCommandType::ShadowAtlasPrePassDirectional] = std::bind(&CRenderManager::ShadowAtlasPrePassDirectional, this, std::placeholders::_1);
```
Comment (AxelSavage) [inline Havtorn/Source/Engine/Graphics/RenderManager.cpp]: Based on the style in this method before this line, binding these RenderFunctions could be moved into its own method?
```suggestion
BindRenderActions();
```

**Review by AxelSavage — APPROVED** [review]
Amazing <3

**Review by AxelSavage — APPROVED** [review]

File: Havtorn/Source/Engine/Graphics/Shaders/SpriteInstanced_VS.hlsl (hunk 11 above)
Comment (AxelSavage) [reply Havtorn/Source/Engine/Graphics/Shaders/SpriteInstanced_VS.hlsl:13]: Seems to me that it's just some diff in indentation length here on git vs in-code-editor :)

**Review by Aki-Gonzalez — APPROVED** [review]
